### PR TITLE
ENH: Geometric Standard Deviation

### LINKED ISSUE #9154: Geometric standard deviation
Small issue, but I wonder if it would make sense to add a geometric standard deviation function (e.g.
 ``scipy.stats.gstd()``) to go along with the geometric mean function (``scipy.stats.gmean()``).  I'd be willing to make a PR if there's interest.

### PR DESCRIPTION
**Pull request by Kai-Striega**
#closes gh-9154

Implementation of the geometric standard deviation function. Handles array_like, ndarrays and masked arrays for multiple dimensions (I may need to add some further tests for edge cases).

To handle invalid values (nans, infty, etc.) and non-positive values `gstd` converts the array to a masked array, in addition to returning a `RuntimeWarning`. This still allows for the geometric sd to be calculated excluding any such values.

@chrisb83 from what I can see you have more experience with stats than myself, would you mind taking a look over this and seeing what can be improved?

### CODE REVIEW

**Review by chrisb83 — COMMENTED** [review]
Hi, 

I had a quick look. I think one might need one Version in mstats that masks invalid input and one version in stats. Otherwise I think this is a useful addition

File: scipy/stats/stats.py [hunk 1]
```diff
@@ -2331,6 +2333,109 @@ def zmap(scores, compare, axis=0, ddof=0):
...
+    """
+
+    if not (isinstance(axis, int) or axis is None):
+        raise ValueError('Invalid axis provided. Axis must be None or int')
+
+    if not isinstance(ddof, int) or ddof < 0:
+        raise ValueError('Invalid ddof provided. A nonnegative int is required')
+
+    try:
+        is_invalid = np.logical_or(np.less_equal(a, 0), ~np.isfinite(a))
```
[the remark this answers is not in the document]
Comment (Kai-Striega) [reply scipy/stats/stats.py]: I've based this off `gmean` which uses masked arrays explicitly. I'll be happy to move some of the sections to `mstats`. I haven't done much (any) work in the stats modules so would you be able to advise on how I should be doing this?

File: scipy/stats/stats.py [hunk 2]
```diff
@@ -2331,6 +2333,109 @@ def zmap(scores, compare, axis=0, ddof=0):
...
+            [17, 18, 19, 20],
+            [21, 22, 23, 24]]])
+    >>> gstd(a, axis=None)
+    2.2944076136018947
+    >>> gstd(a, axis=2)
+    array([[1.82424757, 1.22436866, 1.13183117],
+           [1.09348306, 1.07244798, 1.05914985]])
+    """
+
+    if not (isinstance(axis, int) or axis is None):
```
[the remark this answers is not in the document]
Comment (Kai-Striega) [reply scipy/stats/stats.py]: I've just checked it. `np.std` will raise an `IndexError` for `axis`, this check can be removed. For `ddof` a `RuntimeWarning` is returned with values of `inf` for the standard deviation. In this case I think raising an error would be better.

File: scipy/stats/stats.py [hunk 3]
```diff
@@ -2331,6 +2333,109 @@ def zmap(scores, compare, axis=0, ddof=0):
...
+    Notes
+    -----
+    The geometric standard deviation is sometimes confused with the exponent of
+    the standard deviation of the arithmetic mean, ``exp(std(a))``. Instead the
+    geometric standard deviation is ``exp(std(log(a)))``.
+
+    The default value for `ddof` is different to the default (0) used by other
+    ddof containing functions, such as ``np.std`` and ``np.nanstd``.
+
+    TODO: List "naive" implementation? May encounter numerical errors in some
```
[the remark this answers is not in the document]
Comment (Kai-Striega) [reply scipy/stats/stats.py]: Agreed, I've removed the Todo.

**Comment by ilayn** [comment]
@Kai-Striega  Try not to open branches on the SciPy repo itself but do it on your own fork when sending PRs. I did the same a few days ago.

File: scipy/stats/mstats_basic.py [hunk 4]
```diff
@@ -2218,6 +2219,88 @@ def kurtosis(a, axis=0, fisher=True, bias=True):
...
+    """Calculate the geometric standard deviation of an array
+
+    The geometric standard deviations is defined as the exponent of the
+    standard deviation of ``log(a)``. Mathematically the population
+    geometric standard deviation is evaluated as::
+
+        gstd = exp(std(log(a))
+
+    The geometric standard deviation describes the spread of a set of numbers
+    where :func: `gmean` is the preferred mean. It is a multiplicative
```
Comment (ev-br) [inline scipy/stats/mstats_basic.py]: `:func:` is extraneous here, single backticks do it automatically.

Comment (Kai-Striega) [reply scipy/stats/mstats_basic.py]: Thanks for picking it up, I'll remove it in the next commit.

**Comment by Kai-Striega** [comment]
@ev-br @chrisb83 

The failing test occurs in the masked arrays `less_equal` comparison. This appears to be a bug with numpy's masked arrays (numpy/numpy#4959), not my implementation. 

Do either of you have a recommended fix/workaround for this? I'll be quite busy over the next few weeks but it should be small change and I'd like to get this done.

File: scipy/stats/stats.py [hunk 5]
```diff
@@ -2331,6 +2333,94 @@ def zmap(scores, compare, axis=0, ddof=0):
...
+        Axis along which to operate. Default is 0. If None, compute over
+        the whole array `a`.
+    ddof : int, optional
+        Degrees of freedom correction in the calculation of the
+        geometric standard deviation. Default is 1.
+
+    Returns
+    -------
+    ndarray or float
+        An array of the geometric standard deviation. If `axis` is `1` or `None`
```
Comment (chrisb83) [inline scipy/stats/stats.py]: Suggest to remove backticks around None / 1 to be consistent with notation above

Same in mstats

**Review by chrisb83 — COMMENTED** [review]
For the stats version, one could think about addding an option `ignore_invalid` which, If True, calls the mstats version in case of invalid values, similar to `nan_policy == 'omit'`. Just a suggestion for discussion.

I will review in more detail soon.

**Comment by Kai-Striega** [comment]
@chrisb83 I've added your idea to optionally ignore nans and neatened the implementation. Could you take a look over it and see if there are any further changes before we merge?

File: scipy/stats/stats.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (WarrenWeckesser) [reply scipy/stats/stats.py]: FYI: gmean and hmean have some inconsistencies in their documentation and behavior: https://github.com/scipy/scipy/issues/6551.  As examples, I wouldn't follow them too closely.

**Comment by WarrenWeckesser** [comment]
@chrisb83 wrote:
> For the stats version, one could think about addding an option `ignore_invalid` which, If True, calls the mstats version in case of invalid values, similar to `nan_policy == 'omit'`.

For consistency, the argument that controls how `nan` is handled should be `nan_policy`, just like it is in many other stats functions.  Let's not introduce a different convention in the API.

Is the difference here that you also want to include `inf` in the values that are ignored?

**Comment by Kai-Striega** [comment]
gstd = exp(std(log(A))) so the excluded values will need to be nan, inf, 0 and negatives. I've chosen to go with `include_invalid` for now. If you would like to keep the API consistent I'd be happy to change it.

File: scipy/stats/stats.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kai-Striega) [reply scipy/stats/stats.py]: Thanks for pointing that out, gstd *should* handle tuples axis arguments. I'll take a look at updating the documentation / adding a docstring example tomorrow.

**Comment by Kai-Striega** [comment]
@WarrenWeckesser @chrisb83 

Could either of you take a look at this? It should be good to merge **except** for the `ignore_invalid` argument being passed.

**Comment by WarrenWeckesser** [comment]
I've though some more about this.  For consistency with how it is used elsewhere, `nan_policy` should control only how `nan`s in the input are handled.   I have some ideas on how other "invalid" values should be handled; I'll add more comments tomorrow.

File: scipy/stats/stats.py [hunk 6]
```diff
@@ -2331,6 +2333,107 @@ def zmap(scores, compare, axis=0, ddof=0):
...
+    -------
+    ndarray or float
+        An array of the geometric standard deviation. If `axis` is None or `a`
+        is a 1d array a float is returned.
+
+    Notes
+    -----
+    As the calculation requires use of logarithms the geometric standard
+    deviation only supports strictly positive values. Any non-positive or
+    infinite values will raise a `ValueError` unless ``ignore_invalid=True``.
```
Comment (chrisb83) [inline scipy/stats/stats.py]: TBD if we want to include ignore_invalid

File: scipy/stats/stats.py [hunk 7]
```diff
@@ -2331,6 +2333,107 @@ def zmap(scores, compare, axis=0, ddof=0):
...
+            'safely coerced to any supported types'
+            )
+
+    if all_finite and np.greater(a, 0).all():
+        return np.exp(np.std(np.log(a), axis=axis, ddof=ddof))
+    elif nan_policy == 'omit':
+        return mstats_basic.gstd(a, axis=axis, ddof=ddof)
+    elif nan_policy == 'propagate':
+        return np.nan
+    else:
```
Comment (chrisb83) [inline scipy/stats/stats.py]: you end up here if `nan_policy = 'abc'` I think. You could use _contains_nan in stats/stats.py

File: scipy/stats/stats.py [hunk 8]
```diff
@@ -2331,6 +2333,107 @@ def zmap(scores, compare, axis=0, ddof=0):
...
+    array([[1.82424757, 1.22436866, 1.13183117],
+           [1.09348306, 1.07244798, 1.05914985]])
+    >>> gstd(a, axis=(1,2))
+    array([2.12939215, 1.22120169])
+    """
+    try:
+        with np.errstate(invalid='ignore'):
+            a = np.asanyarray(a)
+            # np.sum avoids creating a huge array in memory
+            all_finite = np.isfinite(a.sum())
```
Comment (chrisb83) [inline scipy/stats/stats.py]: TBD if we need this. -inf is captured by ">0" and +inf probably leads to nan as an output which seems acceptable (?)

**Review by chrisb83 — COMMENTED** [review]
Just minor comments related to handling of invalid input which we still need to agree on. let's wait for warren's feedback

**Comment by WarrenWeckesser** [comment]
Here are several issues dumped into one big honkin' comment... 

-----

Can we avoid the having both `gstd` and `mstats.gstd`?  How hard would it be to make `gstd` accept masked arrays without implementing a separate function in `mstats`?  The following simple function seems to "just work" with masked arrays, but I haven't tested it much (and it doesn't implement 'nan_policy' or provide nice error handling):

    def gstd_naive(x, axis=None, ddof=1):
        return np.exp(np.std(np.log(x), axis=axis, ddof=ddof))

For example, here's `gstd` applied to a regular array:

    In [210]: y = np.array([[2, 8, 32], [10, 20, 30]])

    In [211]: gstd(y, axis=1)
    Out[211]: array([4.        , 1.74289647])

Here's the same data in a masked array; `gstd_naive` computes the correct values:

    In [212]: m = np.ma.masked_array([[2, 8, 32, np.nan], [999, 10, 20, 30]], mask=[[0, 0, 0, 1],[1, 0, 0, 0]])

    In [213]: m
    Out[213]: 
    masked_array(
      data=[[2.0, 8.0, 32.0, --],
            [--, 10.0, 20.0, 30.0]],
      mask=[[False, False, False,  True],
            [ True, False, False, False]],
      fill_value=1e+20)

    In [214]: gstd_naive(m, axis=1)
    Out[214]: 
    masked_array(data=[4.0, 1.7428964697236582],
                 mask=[False, False],
           fill_value=1e+20)


(Personally, I wish the `mstats` module didn't exist, and either we didn't handle masked arrays at all, or treated masked arrays as first class citizens and properly handled them in all our functions.  I tried taking the latter approach in `stats.power_divergence`.  Note that I said "personally"; this is my opinion, not official SciPy policy.)

-----

`nan_policy` must control only how `nan` is handled, and not how negative values, 0 and inf are handled.

-----

Negative values, 0 and inf can be handled in one of the following ways:

1. Quietly return nan.
    Justification: This behavior is documented in the docstring. [Make sure
    it is!] Since it is documented, just be quiet and do it.
2. Return nan and generate an appropriate warning.
    Justification: Sure, it is documented, but who reads the documentation?
    Take an extra step to increase the chances that the user knows they
    got an exceptional result (but don't crash the program).
2. Raise an error.
    Justification: The result is invalid. What is the point of continuing
    the calculation?

SciPy doesn't have a defined policy for this.  Perhaps we should, so when new functions are added, we have clear guidelines for handling input values outside the domains of the function.

I think the most common option among numpy and scipy functions is (2), although in many cases, I suspect the warning is incidental, and not very helpful for the user.  For example,

```
In [154]: from scipy.stats import sem

In [155]: sem([1, 3, 10, np.inf])
/Users/warren/miniconda36/lib/python3.6/site-packages/numpy/core/_methods.py:117: RuntimeWarning: invalid value encountered in subtract
  x = asanyarray(arr - arrmean)
Out[155]: nan
```

It would be much nicer for the user if that warning said something like "sem is not defined for inputs containing inf; returning nan".

An example of (3) is `scipy.stats.hmean`:

```
In [156]: from scipy.stats import hmean

In [157]: hmean([-1, 0, 4, 10])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-157-b6f4e333c4ef> in <module>()
----> 1 hmean([-1, 0, 4, 10])

~/miniconda36/lib/python3.6/site-packages/scipy-1.2.0.dev0+8854df5-py3.6-macosx-10.7-x86_64.egg/scipy/stats/stats.py in hmean(a, axis, dtype)
    380         return size / np.sum(1.0 / a, axis=axis, dtype=dtype)
    381     else:
--> 382         raise ValueError("Harmonic mean only defined if all elements greater "
    383                          "than zero")
    384 

ValueError: Harmonic mean only defined if all elements greater than zero
```

-----

In the following, the return value should not be a masked array:

```
In [94]: y = np.array([[1, 2, 3], [4, 5, np.nan]])

In [95]: gstd(y, axis=1, nan_policy='omit')
Out[95]: 
masked_array(data=[1.7428964697236584, 1.1709159646743776],
             mask=[False, False],
       fill_value=1e+20)
```

-----

It seems like this case should be handled better:

```
In [128]: gstd([10])
/Users/warren/miniconda36/lib/python3.6/site-packages/numpy/core/_methods.py:140: RuntimeWarning: Degrees of freedom <= 0 for slice
  keepdims=keepdims)
/Users/warren/miniconda36/lib/python3.6/site-packages/numpy/core/_methods.py:132: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Out[128]: nan
```

The same issue occurs here (in the second row), but in this case, there is no warning and
the result for the second row is a masked element:

```
In [141]: y = np.array([[2, 8, 32], [10, np.nan, np.nan]])

In [142]: gstd(y, axis=1, nan_policy='omit')
Out[142]: 
masked_array(data=[4.0, --],
             mask=[False,  True],
       fill_value=1e+20)
```

For the above situations, perhaps there should be a warning (or even an error) stating the constraints on `ddof` related to the length of the input.

**Comment by WarrenWeckesser** [comment]
One more comment about the behavior:  when `nan_policy='propagate'` and `axis` is not None, the shape of the result must be the same as when `nan_policy='omit'`.  The value in the returned array is `nan` for each sequence along the given axis that contains `nan`.  For example, if `xn` is

```
In [301]: xn
Out[301]: 
array([[ 1.,  1.,  1., 16.],
       [nan,  1.,  2.,  3.]])
```
then `gstd(xn, axis=1, nan_policy='propagate')` should return `array([4.0, nan])`.  In this case, the return value of `gstd_naive` (defined above) corresponds to the case `nan_policy='propagate'`:

```
In [302]: gstd_naive(xn, axis=1)
Out[302]: array([ 4., nan])
```

File: scipy/stats/stats.py (hunk 7 above)
Comment (Kai-Striega) [reply scipy/stats/stats.py]: The problem with using `_contains_nan` is it -rightly- only checks for nan values. In our case, I wanted to cover *all* values not compatible with ``gstd``.

**Comment by Kai-Striega** [comment]
@WarrenWeckesser I agree with most of what you have said. The original PR had very similar behaviours; handling masked arrays within the same function, checking the sample size, ``ddof`` and invalid values and giving specific messages. It was decided to remove these tests as they were already raising warnings/errors in the NumPy functions. In hindsight, this was probably the wrong decision. The implementation was also only one implementation in stats, which handles masked arrays.

Looking back, I seem to be redoing a lot of the work I had completed before. Ideally, before moving on, I would like to reach a consensus on how this should actually be handled.

I agree, `nan-policy` should be used only for nans only. My proposal would be to use `_contains_nans` to check for nans. Then check for negatives/infinite values, for which we should return an error. GStd is not defined for those values and if the user *really* needs they can always mask those values themselves. 

Regarding the masked array implementation, going back to a single implementation seems the way to go. Do you think that, similar to gmean, gstd should be imported into the `mstats` namespace?

**Comment by chrisb83** [comment]
> SciPy doesn't have a defined policy for this. Perhaps we should, so when new functions are added, we have clear guidelines for handling input values outside the domains of the function.

True, I would say here is it even more complicated since stats functions often have an mstat equivalent. @Kai-Striega sorry for asking to create the mstats version in the beginning. This looked quite straightforward but now you discovered the issue in #9558 in addition to the question how to handle invalid values.

So I see mainly two options:

- only implement a single function in stats, nan_policy only affects nan values and one raises an error / warning if there are invalid (inf, <0) values (i.e. 2 and 3 that Warren suggested)
- one version in stats and one in mstats, use mstats version in case of nans (masking also inf, something that can be fixed with a solution for #9558 later), raise an error / warning for negative values

I think the first option is better, this is also what Kai proposes, I guess.

**Comment by Kai-Striega** [comment]
@chrisb83 I was aiming towards the first option of a single function in stats, nan_policy only affects nan values and an error for invalid values.

@WarrenWeckesser's earlier `gstd_naive` handles seems to handle maksed arrays and nan values well. I do remember there being issues with masked arrays using the standard NumPy functions (numpy/numpy/issues/4959). I'm not sure how, or if, this should be treated.

What I think needs to be done from here:

- [ ] Manage masked arrays in stats and remove the stats version
- [ ] Change `ignore_invalid` to effect only nans inside gstd (leaving #9558 for later)
- [ ] If `ignore_invalid` is `propogate` the shape of the result should be same as with `omit`
- [ ] Raise (more) descriptive error messages in case of invalid values (negatives and infinities)
- [ ] Check for `ddof` with descriptive warning / message

**Comment by Kai-Striega** [comment]
Closing and submitting a new PR #9810.

**Pull request closed without merging**